File: src/claudebox_frontend/src/api/uiState.js
/** UI state API for global and session-scoped preferences via daemon. */

import { workspaceFetch } from './apiClient'

/**
 * Fetch UI state (global and optionally session-scoped).
 * @param {string|null} sessionId - Include session state if provided.
 */
export async function getUiState(sessionId = null) {
  const url = getUrl(sessionId)
  const res = await workspaceFetch(url)
  if (!res.ok) {
    throw new Error('Failed to fetch ui-state')
  }
  return res.json()
}

/** Apply operations to global UI state. */
export function patchGlobalUiState(operations) {
  patchUiState(null, { global: operations })
}

/** Apply operations to session-scoped UI state. */
export function patchSessionUiState(sessionId, operations) {
  patchUiState(sessionId, { session: operations })
}

/**
 * Apply operations to UI state (fire-and-forget).
 * @param {string|null} sessionId
 * @param {Object} data - Operations per scope ({ global, session }).
 */
export function patchUiState(sessionId, data) {
  const url = getUrl(sessionId)
  workspaceFetch(url, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  }).catch(err => {
    // Intentional: persistence is best-effort
    console.warn('uiState: patch failed', err)
  })
}

function getUrl(sessionId) {
  const basePath = '/ui-state'
  return sessionId ? `${basePath}?session_id=${sessionId}` : basePath
}

File: src/claudebox_daemon/handlers/ui_state.py
"""UI state persistence - workspace-scoped HTTP adapters."""

from fastapi import APIRouter

from claudebox import JSONResponse
from ._models import PatchUIStateRequest
from ._shared import WorkspaceDep


router = APIRouter(prefix="/api/workspaces/{workspace_id}")


@router.get("/ui-state")
async def get_ui_state(svc: WorkspaceDep, session_id: str | None = None):
    """Retrieve UI state for global and session-specific data."""

    return JSONResponse(content=svc.ui_state.get(session_id))


@router.patch("/ui-state")
async def patch_ui_state(
    svc: WorkspaceDep,
    body: PatchUIStateRequest,
    session_id: str | None = None,
):
    """Apply patch operations to global and session-specific UI state."""

    return JSONResponse(
        content=svc.ui_state.patch(
            session_id,
            **body.model_dump(by_alias=True, exclude_none=True),
        ),
    )
